File: CS/Reporting_how-to-create-a-label-in-the-pageheader-band-when-a-field-is-dragged-from-the-t457241/T457241/Form1.cs
using DevExpress.XtraReports.Design;
using DevExpress.XtraReports.UI;
using DevExpress.XtraReports.UserDesigner;
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.ComponentModel.Design;
using System.Data;
using System.Drawing;
using System.IO;
using System.Linq;
using System.Runtime.Serialization.Formatters.Binary;
using System.Text;
using System.Windows.Forms;

namespace T457241 {
    public partial class Form1 : DevExpress.XtraEditors.XtraForm {
        ReportDesignTool designer;
        IDesignerHost host;
        public Form1() {
            InitializeComponent();
        }

        private void simpleButton1_Click(object sender, EventArgs e) {
            designer = new ReportDesignTool(new XtraReport1());
            designer.DesignRibbonForm.DesignMdiController.DesignPanelLoaded += DesignMdiController_DesignPanelLoaded;
            designer.ShowRibbonDesignerDialog();
        }

        private void DesignMdiController_DesignPanelLoaded(object sender, DesignerLoadedEventArgs e) {
            if(radioGroup1.SelectedIndex == 0)
                (sender as XRDesignPanel).ComponentAdded += DesignPanel_ComponentAdded;
            else {
                e.DesignerHost.RemoveService(typeof(IFieldListDragDropService));
                e.DesignerHost.AddService(typeof(IFieldListDragDropService), new CustomFieldListDragDropService(e.DesignerHost, (XRDesignPanel)sender));
            }
        }

        private void DesignPanel_ComponentAdded(object sender, System.ComponentModel.Design.ComponentEventArgs e) {
            PageHeaderBand headerBand;
            host = (IDesignerHost)designer.DesignRibbonForm.DesignMdiController.ActiveDesignPanel.Report.Site.GetService(typeof(IDesignerHost));
            if(!(e.Component is XRLabel) && !(e.Component is XRTable)) return;
            if(e.Component is XRLabel) {
                XRLabel label = e.Component as XRLabel;
                if(label.DataBindings.Count == 0) return;
                if(label.Parent is XRTableRow) return;
                headerBand = CreatePageHeaderBand(designer.Report);
                XRLabel newLabel = CopyLabel(label, headerBand);
                headerBand.Controls.Add(newLabel);
                DesignTool.AddToContainer(host, newLabel);
            }
            if(e.Component is XRTable) {
                headerBand = CreatePageHeaderBand(designer.Report);
                XRTableRow sourceRow = (e.Component as XRTable).Rows[0];
                XRTable headerTable = CreateTable(sourceRow, headerBand.HeightF);
                headerBand.Controls.Add(headerTable);
                DesignTool.AddToContainer(host, headerTable);
            }
        }

        PageHeaderBand CreatePageHeaderBand(XtraReport report) {
            PageHeaderBand headerBand = report.Bands[BandKind.PageHeader] as PageHeaderBand;
            // Remove the comments if you need to recreate the PageHeader band
            //if(headerBand != null) {
            //    for(int i = headerBand.Controls.Count - 1; i >= 0; i--)
            //        DevExpress.XtraReports.Design.DesignTool.RemoveFromContainer(host, headerBand.Controls[i]);
            //    headerBand.Controls.Clear();
            //    DevExpress.XtraReports.Design.DesignTool.RemoveFromContainer(host, headerBand);
            if(headerBand == null) {
                headerBand = new PageHeaderBand();
                headerBand.HeightF = 0;
                headerBand.Visible = true;
                DesignTool.AddToContainer(host, headerBand);
            }
            return headerBand;
        }

        XRTable CreateTable(XRTableRow sourceRow, float height) {
            XRTable table = new XRTable();
            XRTableRow headerTableRow = new XRTableRow();
            table.BeginInit();
            table.Rows.Add(headerTableRow);
            foreach(XRTableCell cell in sourceRow.Cells)
                CopyCell(cell, headerTableRow);
            table.Borders = DevExpress.XtraPrinting.BorderSide.All;
            table.LocationF = new PointF(0, height);
            table.AdjustSize();
            table.EndInit();
            return table;
        }

        XRLabel CopyLabel(XRLabel label, PageHeaderBand band) {
            string text = "XRLabel";
            XtraReport report = band.Report as XtraReport;
            XRLabel newLabel = new XRLabel();
            newLabel.WidthF = label.WidthF;
            newLabel.HeightF = label.HeightF;
            newLabel.BackColor = Color.Green;
            newLabel.ForeColor = Color.Yellow;
            newLabel.TextAlignment = DevExpress.XtraPrinting.TextAlignment.MiddleCenter;
            newLabel.Font = new Font("Calibry", 11, FontStyle.Bold);
            newLabel.Borders = DevExpress.XtraPrinting.BorderSide.All;
            newLabel.LocationF = new PointF(0, band.HeightF);
            newLabel.WidthF = report.PageWidth - report.Margins.Left - report.Margins.Right;
            if(label.DataBindings.Count > 0) {
                text = label.DataBindings[0].DataMember;
                int index = text.LastIndexOf(".");
                if(index > 0)
                    text = text.Substring(index + 1);
            }
            newLabel.Text = text;
            return newLabel;
        }

        void CopyCell(XRTableCell source, XRTableRow row) {
            XRTableCell cell = new XRTableCell();
            cell.WidthF = source.WidthF;
            cell.HeightF = source.HeightF;
            cell.BackColor = Color.Green;
            cell.ForeColor = Color.Yellow;
            cell.TextAlignment = DevExpress.XtraPrinting.TextAlignment.MiddleCenter;
            cell.Font = new Font("Calibry", 11, FontStyle.Bold);
            string text = source.DataBindings[0].DataMember;
            int index = text.LastIndexOf(".");
            if(index > 0)
                text = text.Substring(index + 1);
            cell.Text = text;
            row.Cells.Add(cell);
            DesignTool.AddToContainer(host, cell);
        }
    }
}

File: VB/Reporting_how-to-create-a-label-in-the-pageheader-band-when-a-field-is-dragged-from-the-t457241/T457241/Form1.vb
Imports DevExpress.XtraReports.Design
Imports DevExpress.XtraReports.UI
Imports DevExpress.XtraReports.UserDesigner
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.ComponentModel.Design
Imports System.Data
Imports System.Drawing
Imports System.IO
Imports System.Linq
Imports System.Runtime.Serialization.Formatters.Binary
Imports System.Text
Imports System.Windows.Forms

Namespace T457241
	Partial Public Class Form1
		Inherits DevExpress.XtraEditors.XtraForm

		Private designer As ReportDesignTool
		Private host As IDesignerHost
		Public Sub New()
			InitializeComponent()
		End Sub

		Private Sub simpleButton1_Click(ByVal sender As Object, ByVal e As EventArgs) Handles simpleButton1.Click
			designer = New ReportDesignTool(New XtraReport1())
			AddHandler designer.DesignRibbonForm.DesignMdiController.DesignPanelLoaded, AddressOf DesignMdiController_DesignPanelLoaded
			designer.ShowRibbonDesignerDialog()
		End Sub

		Private Sub DesignMdiController_DesignPanelLoaded(ByVal sender As Object, ByVal e As DesignerLoadedEventArgs)
			If radioGroup1.SelectedIndex = 0 Then
				AddHandler TryCast(sender, XRDesignPanel).ComponentAdded, AddressOf DesignPanel_ComponentAdded
			Else
				e.DesignerHost.RemoveService(GetType(IFieldListDragDropService))
				e.DesignerHost.AddService(GetType(IFieldListDragDropService), New CustomFieldListDragDropService(e.DesignerHost, DirectCast(sender, XRDesignPanel)))
			End If
		End Sub

		Private Sub DesignPanel_ComponentAdded(ByVal sender As Object, ByVal e As System.ComponentModel.Design.ComponentEventArgs)
			Dim headerBand As PageHeaderBand
			host = DirectCast(designer.DesignRibbonForm.DesignMdiController.ActiveDesignPanel.Report.Site.GetService(GetType(IDesignerHost)), IDesignerHost)
			If Not(TypeOf e.Component Is XRLabel) AndAlso Not(TypeOf e.Component Is XRTable) Then
				Return
			End If
			If TypeOf e.Component Is XRLabel Then
				Dim label As XRLabel = TryCast(e.Component, XRLabel)
				If label.ExpressionBindings.Count = 0 Then
					Return
				End If
				If TypeOf label.Parent Is XRTableRow Then
					Return
				End If
				headerBand = CreatePageHeaderBand(designer.Report)
				Dim newLabel As XRLabel = CopyLabel(label, headerBand)
				headerBand.Controls.Add(newLabel)
				DesignToolHelper.AddToContainer(host, newLabel)
			End If
			If TypeOf e.Component Is XRTable Then
				headerBand = CreatePageHeaderBand(designer.Report)
				Dim sourceRow As XRTableRow = (TryCast(e.Component, XRTable)).Rows(0)
				Dim headerTable As XRTable = CreateTable(sourceRow, headerBand.HeightF)
				headerBand.Controls.Add(headerTable)
				DesignToolHelper.AddToContainer(host, headerTable)
			End If
		End Sub

		Private Function CreatePageHeaderBand(ByVal report As XtraReport) As PageHeaderBand
			Dim headerBand As PageHeaderBand = TryCast(report.Bands(BandKind.PageHeader), PageHeaderBand)
			' Remove the comments if you need to recreate the PageHeader band
			'if(headerBand != null) {
			'    for(int i = headerBand.Controls.Count - 1; i >= 0; i--)
			'        DesignToolHelper.RemoveFromContainer(host, headerBand.Controls[i]);
			'    headerBand.Controls.Clear();
			'    DesignToolHelper.RemoveFromContainer(host, headerBand);
			If headerBand Is Nothing Then
				headerBand = New PageHeaderBand()
				headerBand.HeightF = 0
				headerBand.Visible = True
				DesignToolHelper.AddToContainer(host, headerBand)
			End If
			Return headerBand
		End Function

		Private Function CreateTable(ByVal sourceRow As XRTableRow, ByVal height As Single) As XRTable
			Dim table As New XRTable()
			Dim headerTableRow As New XRTableRow()
			table.BeginInit()
			table.Rows.Add(headerTableRow)
			For Each cell As XRTableCell In sourceRow.Cells
				CopyCell(cell, headerTableRow)
			Next cell
			table.Borders = DevExpress.XtraPrinting.BorderSide.All
			table.LocationF = New PointF(0, height)
			table.AdjustSize()
			table.EndInit()
			Return table
		End Function

		Private Function CopyLabel(ByVal label As XRLabel, ByVal band As PageHeaderBand) As XRLabel
            Dim labelText As String = "XRLabel"
            Dim report As XtraReport = TryCast(band.Report, XtraReport)
			Dim newLabel As New XRLabel()
			newLabel.LocationF = label.LocationF
			newLabel.WidthF = label.WidthF
            newLabel.HeightF = label.HeightF
            newLabel.BackColor = Color.Green
            newLabel.ForeColor = Color.Yellow
            newLabel.TextAlignment = DevExpress.XtraPrinting.TextAlignment.MiddleCenter
            newLabel.Font = New Font("Calibry", 11, FontStyle.Bold)
			newLabel.Borders = DevExpress.XtraPrinting.BorderSide.All
			If label.ExpressionBindings.Count > 0 Then
				labelText = label.ExpressionBindings(0).Expression
				Dim index As Integer = labelText.LastIndexOf(".")
				If index > 0 Then
					labelText = labelText.Substring(index + 1)
				End If
			End If
			newLabel.Text = labelText
			Return newLabel
		End Function

		Private Sub CopyCell(ByVal source As XRTableCell, ByVal row As XRTableRow)
			Dim cell As New XRTableCell()
			cell.WidthF = source.WidthF
			cell.HeightF = source.HeightF
			cell.BackColor = Color.Green
			cell.ForeColor = Color.Yellow
			cell.TextAlignment = DevExpress.XtraPrinting.TextAlignment.MiddleCenter
			cell.Font = New Font("Calibry", 11, FontStyle.Bold)
			Dim cellText As String = source.ExpressionBindings(0).Expression
			Dim index As Integer = cellText.LastIndexOf(".")
            If index > 0 Then
                cellText = cellText.Substring(index + 1)
            End If
            cell.Text = cellText
			row.Cells.Add(cell)
			DesignToolHelper.AddToContainer(host, cell)
		End Sub
	End Class
End Namespace
